### PR TITLE
[ADD] product_attribute_value_combination_conditions

### PR DESCRIPTION
**Pull request by mikelarre**
product_attribute_value_combination_conditions

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Error PEP8:

```
./product_attribute_value_combination_conditions/models/product.py:7:1: E302 expected 2 blank lines, found 1
```

File: product_attribute_value_combination_conditions/models/product.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    @api.constrains('predecessors', 'initial')
+    def check_initial_has_predecessors(self):
+        if self.initial and self.predecessors:
+            raise exceptions.Warning('initial attribute cant content '
+                                     'predecessors')
+
+    @api.constrains('predecessors', 'predecessors_constraint')
+    def check_predecessors_values(self):
+        if self.predecessors_constraint - self.possible_predecessor_values:
+            raise exceptions.Warning('There are values that not longer exist')
```
Comment (pedrobaeza) [inline product_attribute_value_combination_conditions/models/product.py]: Cambiar por "There are values in possibles values that don't belong anymore to the predecessors"

**Comment by pedrobaeza** [comment]
Cambios en interfaz:
- Cuando initial está marcado, hay que deshabilitar también predecessors constraints
- Value constraints debe ir después de product attribute value

File: product_attribute_value_combination_conditions/models/product.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            self.possible_predecessor_values = values
+
+    initial = fields.Boolean(string='Initial')
+    predecessors = fields.Many2many(
+        comodel_name='product.attribute',
+        relation='attribute_predecessors_rel',
+        column1='attribute_line',
+        column2='product_attribute',
+        domain="[('id', 'in', possible_values[0][2])]")
+    possible_values = fields.Many2many(
```
Comment (pedrobaeza) [inline product_attribute_value_combination_conditions/models/product.py]: Renómbralo a `possible_predecessors` para ser más intuitivo

File: product_attribute_value_combination_conditions/models/product.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                    'attribute_id'):
+                raise exceptions.Warning('There is a circular dependency')
+
+
+class ProductAttributeLine(models.Model):
+    _inherit = "product.attribute.line"
+
+    @api.one
+    @api.depends('product_tmpl_id.attribute_line_ids')
+    def _get_possible_attributes(self):
```
Comment (pedrobaeza) [inline product_attribute_value_combination_conditions/models/product.py]: renombrar a `_get_possible_predecessors` para ser más intuitivo

**Comment by matt454357** [comment]
I don't understand the concept of predecessors.  How is the predecessor used to constrain attribute values?  I would like to test and give feedback.

**Comment by matt454357** [comment]
Can you give an example of the situation you are targeting?

I am testing with a product template having the following attributes and values.
- Engine
  - 129kW
  - 151kW
  - 194kW
- Gearbox
  - 12kNm
  - 16kNm

I want to apply a constraint that dis-allows (blocks) the combination of the 194kW engine and the 12kNm gearbox.  In other words, I will only allow the 16kNm gearbox when the user chooses the 194kW engine.

**Comment by anajuaristi** [comment]
So in the example you give... you should set 
engine 194kw as precedent of gearbox 16knm 

On the other side you would set engine 129kW and engine 151kW as precedents of Gearbox 12kNm

**Comment by pedrobaeza** [comment]
@anajuaristi, no, that's not the meaning for the module. A predecessor is the attribute that has to be set before you can set the one you're defining. In the case you should create:
- Attribute: `Engine`. Predeccesors: None.
- Attribute: `Gearbox`. Predeccesors: `Engine`.

This way, you assure that the user set `Engine` attribute (it is presented at the beginning) before presenting/constraining the attribute `Gearbox`.

And for the restriction you want to apply, you have the "Value constraints" list, where you have to set 2 lines:
- Predeccesors values: `129kW`, `151kW`. Allowed values: `12kNm`, `16kNm`
- Predeccesors values: `194kW`. Allowed values: `16kNm`

**Pull request closed without merging**